AP_DDS: Added ROS 2 service support

arshPratap opened the pull request:
This is work in progress PR that aims to provide support for service support as requested in #23822 and will work as a foundation for #23363 . Currently we focus on providing support for Arming / Disarming motors.

@Ryanf55 @srmainwaring @tridge would like to know your thoughts on this

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -473,6 +520,9 @@ bool AP_DDS_Client::init()
     uxr_set_topic_callback(&session, AP_DDS_Client::on_topic, &count);
+    
+    // ROS-2 Service : Register service request callbacks
+    uxr_set_request_callback(&session, AP_DDS_Client::on_request, 0);
```
What does the last parameter of zero mean? Can you make it a constexpr variable?

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -598,6 +648,44 @@ bool AP_DDS_Client::create()
+            requests[0] = replier_req_id;
+
+            if (!uxr_run_session_until_all_status(&session, requestTimeoutMs, requests, status, nRequests)) {
+                GCS_SEND_TEXT(MAV_SEVERITY_ERROR,"XRCE Client: Service/Replier session request failure for index '%d'",(int)i);
```
Why do you cast `i` to `int` when you can just use the size_t unsigned print formatter? 
`printf("%zu\n", x);`

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.h:
```diff
@@ -187,7 +188,20 @@ class AP_DDS_Client
-
+    //! @brief Convenience grouping for a single "channel" of services
+    struct Service_table{
+        //req_id
```
What are the purpose of the comments that repeat the variable name?
Feel free to use doxygen comments that explain a bit more like so:
`//! @brief The request ID for the service`

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Service_Table.h:
```diff
@@ -0,0 +1,21 @@
+        .req_profile_label = "",
+        .rep_profile_label = "ArmMotors_Replier",
+    }
+};
```
Missing newline at the end of this file.

Ryanf55 commented on libraries/AP_DDS/dds_xrce_profile.xml:
```diff
@@ -5,6 +5,8 @@
       <name>Ardupilot_DDS_XRCE_Client</name>
     </rtps>
   </participant>
+  <replier profile_name="ArmMotors_Replier" service_name = "ArmMotorsService" request_type = "ArmMotors_Request" reply_type = "ArmMotors_Response">
```
These look nice and easy to add :D

Ryanf55 reviewed (commented):
Looks good so far. Could you update the README with how to use it so I can test it out? 

Are there not any new messages to add to the build?

arshPratap commented:
> Looks good so far. Could you update the README with how to use it so I can test it out?
> 
> Are there not any new messages to add to the build?

@Ryanf55 thanks for the suggestions , have filled in the README as requested ... I hope you guys are able to successfully setup the Integration Services ( just in case , you can also take a look at [this](https://github.com/arshPratap/ROS2_DDS_Service_Demo/tree/master) to get yourself a bit more familiar with this aspect of the project)
Currently on my way to working on the other requested changes
Thanks a lot for your quick response

arshPratap commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -433,6 +437,49 @@ void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, ui
+
+    if(arm == true) {
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO,"DDS Client: Request for arming recieved");
+    } else {
```
Resolved @Ryanf55

arshPratap commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -433,6 +437,49 @@ void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, ui
+    (void) object_id; (void) request_id; (void) length; (void) args;
+    
+    bool arm;
+    ucdr_deserialize_bool(ub,&arm);
```
Thanks for this..have added it accordingly

arshPratap commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -433,6 +437,49 @@ void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, ui
+    bool result = arm;
+
+    ucdr_init_buffer(&reply_ub, reply_buffer, sizeof(reply_buffer));
+    ucdr_serialize_bool(&reply_ub,result);
```
Done

arshPratap commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -19,6 +19,10 @@ static constexpr uint16_t DELAY_LOCAL_VELOCITY_TOPIC_MS = 33;
 static constexpr uint16_t DELAY_CLOCK_TOPIC_MS = 10;
 static char WGS_84_FRAME_ID[] = "WGS-84";
+
+static uxrStreamId reliable_in;
```
`on_request` function is a static member function and it causes some linking error when the said variables are defined as static members , hence have assigned them here

tridge replied:
fixed to use trampolines so non-static

arshPratap commented:
@Ryanf55 @srmainwaring have rebased the branch to the latest master branch  and have merged @srmainwaring 's PRs into this branch (as requested). @Ryanf55 I have also worked upon all the suggested changes and would love to know your thoughts on them

srmainwaring commented:
Thanks @arshPratap - the commits to `AP_DDS` and `Tools` need to be separated which is why they were broken out in the additions I suggested. Within each of these libraries we could break the changes into something like:

1. AP_DDS: add ROS 2 arming service
2. Tools: add ROS 2 custom messages and services
3. Tools: add ROS 2 launch for eProsima integration service

and squash the changes within those groupings.

Alternatively we could break them out into 3 separate PRs - with the custom messages [2] going in first, then [1] which depends on [2], and finally [3] which is intended to simplify the process of setting up and running the service example.  @Ryanf55 - your thoughts?

Ryanf55 commented:
> Thanks @arshPratap - the commits to `AP_DDS` and `Tools` need to be separated which is why they were broken out in the additions I suggested. Within each of these libraries we could break the changes into something like:
> 
> 1. AP_DDS: add ROS 2 arming service
> 2. Tools: add ROS 2 custom messages and services
> 3. Tools: add ROS 2 launch for eProsima integration service
> 
> and squash the changes within those groupings.
> 
> Alternatively we could break them out into 3 separate PRs - with the custom messages [2] going in first, then [1] which depends on [2], and finally [3] which is intended to simplify the process of setting up and running the service example. @Ryanf55 - your thoughts?
> 
> Did you manage to resolve the error reported in the integration service terminal that is raised when the service returns it's reply message? I see this in macOS and Ubuntu (so not a macOS thing).
> 
> **Update**:
> 
> This is the error message I see in the integration service terminal:
> 
> ```
> [Integration Service][INFO] [is::sh::ROS2] Receiving request from ROS 2 for service request topic 'arm_motors_Request'
> [Integration Service][INFO] [is::sh::FastDDS::Server] Translating request from Integration Service to DDS for service request topic 'ArmMotorsService_Request': [[ Structure: <ap_custom_services/ArmMotors:request>
>     [arm] <bool>  1
>  ]]
> 2023-06-28 00:29:48.672 [RTPS_READER_HISTORY Error] Change payload size of '12' bytes is larger than the history payload size of '8' bytes and cannot be resized. -> Function can_change_be_added_nts
> Request timeout.
> ```

Either approach of 3 commits or 3 PR's sounds good to me, whichever works best for Arsh. I agree- it cannot be done in one single commit.

srmainwaring commented:
@arshPratap - here are the logs from running the Arming_DDS_Server example. I see the same error in the IS terminal when running in SITL.


MicroXRCEAgent:

```bash
$ MicroXRCEAgent udp4 -p 2019 -v6
[1688730229.983349] info     | UDPv4AgentLinux.cpp | init                     | running...             | port: 2019
[1688730229.984016] info     | Root.cpp           | set_verbose_level        | logger setup           | verbose_level: 6
[1688730263.205322] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0x00000000, len: 24, data: 
0000: 80 00 00 00 00 01 10 00 58 52 43 45 01 00 01 0F CC CC DD DD 81 00 FC 01
[1688730263.205465] info     | Root.cpp           | create_client            | create                 | client_key: 0xCCCCDDDD, session_id: 0x81
[1688730263.205503] info     | SessionManager.hpp | establish_session        | session established    | client_key: 0xCCCCDDDD, address: 127.0.0.1:4073
[1688730263.205556] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 19, data: 
0000: 81 00 00 00 04 01 0B 00 00 00 58 52 43 45 01 00 01 0F 00
[1688730263.205634] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0xCCCCDDDD, len: 280, data: 
0000: 81 80 00 00 01 05 60 00 00 0A 00 11 01 02 00 42 51 00 00 00 3C 64 64 73 3E 3C 70 61 72 74 69 63
0020: 69 70 61 6E 74 3E 3C 72 74 70 73 3E 3C 6E 61 6D 65 3E 41 72 6D 4D 6F 74 6F 72 73 5F 53 65 72 76
0040: 65 72 3C 2F 6E 61 6D 65 3E 3C 2F 72 74 70 73 3E 3C 2F 70 61 72 74 69 63 69 70 61 6E 74 3E 3C 2F
0060: 64 64 73 3E 00 8B 00 00 01 05 AC 00 00 0B 00 18 08 02 00 00 9E 00 00 00 3C 64 64 73 3E 3C 72 65
0080: 70 6C 69 65 72 20 70 72 6F 66 69 6C 65 5F 6E 61 6D 65 3D 22 41 72 6D 4D 6F 74 6F 72 73 5F 52 65
00A0: 70 6C 69 65 72 22 73 65 72 76 69 63 65 5F 6E 61 6D 65 3D 22 41 72 6D 4D 6F 74 6F 72 73 53 65 72
00C0: 76 69 63 65 22 72 65 71 75 65 73 74 5F 74 79 70 65 3D 22 41 72 6D 4D 6F 74 6F 72 73 5F 52 65 71
00E0: 75 65 73 74 22 72 65 70 6C 79 5F 74 79 70 65 3D 22 41 72 6D 4D 6F 74 6F 72 73 5F 52 65 73 70 6F
0100: 6E 73 65 22 3E 3C 2F 72 65 70 6C 69 65 72 3E 3C 2F 64 64 73 3E 00 00 11
[1688730263.218969] info     | ProxyClient.cpp    | create_participant       | participant created    | client_key: 0xCCCCDDDD, participant_id: 0x001(1)
[1688730263.219024] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 14, data: 
0000: 81 80 00 00 05 01 06 00 00 0A 00 11 00 00
[1688730263.219195] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0xCCCCDDDD, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 01 00 00 00 80
[1688730263.219869] info     | ProxyClient.cpp    | create_replier           | replier created        | client_key: 0xCCCCDDDD, requester_id: 0x001(7), participant_id: 0x001(1)
[1688730263.219902] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 14, data: 
0000: 81 80 01 00 05 01 06 00 00 0B 00 18 00 00
[1688730263.219916] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 01 00 00 00 80
[1688730263.219951] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0xCCCCDDDD, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 02 00 00 00 80
[1688730263.219969] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0xCCCCDDDD, len: 24, data: 
0000: 81 80 01 00 08 01 10 00 00 0C 00 18 80 00 00 01 FF FF 00 00 00 00 00 00
[1688730263.220060] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 02 00 00 00 80
[1688730422.636606] debug    | Replier.cpp        | read_fn                  | [==>> DDS <<==]        | client_key: 0x00000001, len: 28, data: 
0000: 01 0F 57 FA 62 F2 8D 42 01 00 00 00 00 00 01 03 00 00 00 00 01 00 00 00 01 00 00 00
[1688730422.636692] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 40, data: 
0000: 81 80 02 00 09 01 20 00 00 0C 00 18 01 0F 57 FA 62 F2 8D 42 01 00 00 00 00 00 01 03 00 00 00 00
0020: 01 00 00 00 01 00 00 00
[1688730422.636784] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0xCCCCDDDD, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 03 00 00 00 80
[1688730423.636880] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0xCCCCDDDD, len: 44, data: 
0000: 81 80 02 00 07 01 24 00 00 0D 00 18 01 0F 57 FA 62 F2 8D 42 01 00 00 00 00 00 01 03 00 00 00 00
0020: 01 00 00 00 01 00 00 00 00 00 00 00
[1688730423.637018] debug    | Replier.cpp        | write                    | [** <<DDS>> **]        | client_key: 0x00000001, len: 32, data: 
0000: 01 0F 57 FA 62 F2 8D 42 01 00 00 00 00 00 01 03 00 00 00 00 01 00 00 00 01 00 00 00 00 00 00 00
[1688730423.637115] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 03 00 00 00 80
```

Arming_DDS_Server:

```bash
$ ./a.out 127.0.0.1 2019 1
Entities Created Successfully 
Request recieved for Arming 
Reply : Armed 
```

Integration Service

```bash
$ LD_LIBRARY_PATH=/Users/rhys/Code/ros2/xrce-dds/eprosima_ws/install/is-fastdds/lib integration-service ./src/ROS2_DDS_Service_Demo/AP_Services/Arming_DDS_Server/Arm_Motors_DDS_IS_config.yaml --is-prefix-path /Users/rhys/Code/ros2/xrce-dds/eprosima_ws/install/is-ros2/lib/is/ros2 --is-prefix-path /Users/rhys/Code/ros2/xrce-dds/eprosima_ws/install/is-ros2-mix-generator/lib/is
[Integration Service][WARN] [is::sh::FastDDS] Participant not provided in configuration file. A participant using the default transport locators and Domain ID 0 will be created.
[Integration Service][INFO] [is::sh::FastDDS::Participant] Created Fast DDS participant 'default_IS-FastDDS-SH_participant' with default QoS attributes and Domain ID: 0
[Integration Service][INFO] [is::sh::FastDDS] Configured!
[Integration Service][INFO] [is::sh::ROS2] Created node '/is_ros2_node_16807' with Domain ID: 0
[Integration Service][INFO] [is::sh::ROS2] Configured!
[Integration Service][INFO] [is::sh::FastDDS::Server] Publisher for topic 'ArmMotorsService_Request' matched
[Integration Service][INFO] [is::sh::FastDDS::Server] Subscriber for topic 'ArmMotorsService_Reply' matched
[Integration Service][INFO] [is::sh::FastDDS] Server created for service 'ArmMotorsService', with request_type 'ArmMotors_Request' and reply_type 'ArmMotors_Response'
[Integration Service][INFO] [is::core::Config]  [dds SystemHandle] Produced a service provider for the service 'arm_motors', with request type 'ArmMotors_Request' and reply type 'ArmMotors_Response'.
[Integration Service][INFO] [is::sh::ROS2] Creating entity using the default QoS.
[Integration Service][INFO] [is::sh::ROS2] Created service client for service 'arm_motors' with type 'ap_custom_services/ArmMotors:response' on node '/is_ros2_node_16807'
[Integration Service][INFO] [is::core::Config] [ros2 SystemHandle] Produced a service client for the service 'arm_motors', with request type 'ArmMotors_Request' and reply type 'ArmMotors_Response'.
[Integration Service][INFO] [is::sh::ROS2] Receiving request from ROS 2 for service request topic 'arm_motors_Request'
[Integration Service][INFO] [is::sh::FastDDS::Server] Translating request from Integration Service to DDS for service request topic 'ArmMotorsService_Request': [[ Structure: <ap_custom_services/ArmMotors:request>
    [arm] <bool>  1
 ]]
2023-07-07 12:47:03.637 [RTPS_READER_HISTORY Error] Change payload size of '12' bytes is larger than the history payload size of '8' bytes and cannot be resized. -> Function can_change_be_added_nts
Request timeout.
```

Arm Motors Client:

```bash
$ ros2 run ap_service_clients arm_motors_client 1
[INFO] [1688730369.125680900] [rclcpp]: Enter Arming Command
arm
[INFO] [1688730427.639468326] [rclcpp]: Motors : Disarmed Successfully
```

srmainwaring commented:
@arshPratap - additional tools: https://github.com/srmainwaring/ROS2_DDS_Service_Demo/tree/wips/wip-ros2-arming-server-fastdds

#### Build:

Build a FastDDS server and client based on the AddTwoInts example, and the Micro XRCE Client example service using cmake:

```bash
$ cd ~/ws/src/ROS2_DDS_Service_Demo/AP_Services/Arming_DDS_Server
$ mkdir build && cd build
$ cmake ..
$ make
```

Expected build outputs:

- `ArmMotors` the FastDDS server / client
- `ArmMotorsMicro` the Micro XRCE Client arming service (previously built directly and referred to as `a.out` above)

### Example 1: ROS 2 <=> FastDDS:

- Run an arming service using FastDDS (i.e. no Micro XRCE Agent invoked).
- Run the Integration Service to translate from ROS 2 to FastDDS.
- Run a service call using ROS 2 CLI.

#### Run

Integration Service:

```bash
cd ~ws
$ LD_LIBRARY_PATH=$(pwd)/install/is-fastdds/lib integration-service $(pwd)/src/ROS2_DDS_Service_Demo/AP_Services/Arming_DDS_Server/Arm_Motors_DDS_IS_config.yaml --is-prefix-path $(pwd)/install/is-ros2/lib/is/ros2 --is-prefix-path $(pwd)/install/is-ros2-mix-generator/lib/is

[Integration Service][WARN] [is::sh::FastDDS] Participant not provided in configuration file. A participant using the default transport locators and Domain ID 0 will be created.
[Integration Service][INFO] [is::sh::FastDDS::Participant] Created Fast DDS participant 'default_IS-FastDDS-SH_participant' with default QoS attributes and Domain ID: 0
[Integration Service][INFO] [is::sh::FastDDS] Configured!
[Integration Service][INFO] [is::sh::ROS2] Created node '/is_ros2_node_16807' with Domain ID: 0
[Integration Service][INFO] [is::sh::ROS2] Configured!
[Integration Service][INFO] [is::sh::FastDDS] Server created for service 'ArmMotorsService', with request_type 'ArmMotors_Request' and reply_type 'ArmMotors_Response'
[Integration Service][INFO] [is::core::Config]  [dds SystemHandle] Produced a service provider for the service 'arm_motors', with request type 'ArmMotors_Request' and reply type 'ArmMotors_Response'.
[Integration Service][INFO] [is::sh::ROS2] Creating entity using the default QoS.
[Integration Service][INFO] [is::sh::ROS2] Created service client for service 'arm_motors' with type 'ap_custom_services/ArmMotors:response' on node '/is_ros2_node_16807'
[Integration Service][INFO] [is::core::Config] [ros2 SystemHandle] Produced a service client for the service 'arm_motors', with request type 'ArmMotors_Request' and reply type 'ArmMotors_Response'.
[Integration Service][INFO] [is::sh::FastDDS::Server] Publisher for topic 'ArmMotorsService_Request' matched
[Integration Service][INFO] [is::sh::FastDDS::Server] Subscriber for topic 'ArmMotorsService_Reply' matched
```

FastDDS arming service:

```bash
$ cd ~/ws/src/ROS2_DDS_Service_Demo/AP_Services/Arming_DDS_Server/build
$ ./ArmMotors -m server

ArmMotorsService server running under DDS Domain ID: 0
ArmMotorsService server running. Please press enter to stop it at any time.
```

ROS 2 CLI service call:

```bash
ros2 service call /arm_motors ap_custom_services/srv/ArmMotors "{arm: 1}"
requester: making request: ap_custom_services.srv.ArmMotors_Request(arm=True)

response:
ap_custom_services.srv.ArmMotors_Response(result=True)
```

#### Outputs

Integration Service output:

```bash
[Integration Service][INFO] [is::sh::ROS2] Receiving request from ROS 2 for service request topic 'arm_motors_Request'
[Integration Service][INFO] [is::sh::FastDDS::Server] Translating request from Integration Service to DDS for service request topic 'ArmMotorsService_Request': [[ Structure: <ap_custom_services/ArmMotors:request>
    [arm] <bool>  1
 ]]
[Integration Service][INFO] [is::sh::FastDDS::Server] Receiving reply from DDS for service reply topic 'ArmMotorsService_Reply'
[Integration Service][INFO] [is::sh::ROS2] Translating reply from Integration Service to ROS 2 for service reply topic 'arm_motors_Reply': [[ Structure: <ArmMotors_Response>
    [result] <bool>  1
 ]]
```

FastDDS arming service:

```bash
ArmMotorsService server: processing incoming request, arm 1
	 - Sending back response: 1
```


### Example 2: FastDDS <=> Micro XRCE Agent <=> Micro XRCE Client:

- Run an arming service using the Micro XRCE Client API.
- Run the Micro XRCE Agent using udp4 on port 2019.
- Run the integration service to translate from ROS 2 or FastDDS to FastDDS.
- Run a service call using FastDDS or the ROS 2 CLI.

#### Run

Integration Service:

(As per the first example).

Micro XRCE Agent:

```bash]
MicroXRCEAgent udp4 -p 2019 -v6
[1688828954.031760] info     | UDPv4AgentLinux.cpp | init                     | running...             | port: 2019
[1688828954.032428] info     | Root.cpp           | set_verbose_level        | logger setup           | verbose_level: 6
```

Micro XRCE Client arming service:

```bash
$ cd ~/ws/src/ROS2_DDS_Service_Demo/AP_Services/Arming_DDS_Server/build
./ArmMotorsMicro 127.0.0.1 2019
```

FastDDS service call:

```bash
$ cd ~/ws/src/ROS2_DDS_Service_Demo/AP_Services/Arming_DDS_Server/build
$ ./ArmMotors -m client -c 1
ArmMotorsService client running under DDS Domain ID: 0
ArmMotorsService client performing 1 requests.
ArmMotorsService client:
	 - Request 1
2023-07-08 16:15:11.918 [RTPS_READER_HISTORY Error] Change payload size of '12' bytes is larger than the history payload size of '11' bytes and cannot be resized. -> Function can_change_be_added_nts
	 - Error: client exceeded maximum time to wait for a response!
```

#### Outputs

Integration Service output:

```bash
$ LD_LIBRARY_PATH=$(pwd)/install/is-fastdds/lib integration-service $(pwd)/src/ROS2_DDS_Service_Demo/AP_Services/Arming_DDS_Server/Arm_Motors_DDS_IS_config.yaml --is-prefix-path $(pwd)/install/is-ros2/lib/is/ros2 --is-prefix-path $(pwd)/install/is-ros2-mix-generator/lib/is

[Integration Service][WARN] [is::sh::FastDDS] Participant not provided in configuration file. A participant using the default transport locators and Domain ID 0 will be created.
[Integration Service][INFO] [is::sh::FastDDS::Participant] Created Fast DDS participant 'default_IS-FastDDS-SH_participant' with default QoS attributes and Domain ID: 0
[Integration Service][INFO] [is::sh::FastDDS] Configured!
[Integration Service][INFO] [is::sh::ROS2] Created node '/is_ros2_node_16807' with Domain ID: 0
[Integration Service][INFO] [is::sh::ROS2] Configured!
[Integration Service][INFO] [is::sh::FastDDS::Server] Publisher for topic 'ArmMotorsService_Request' matched
[Integration Service][INFO] [is::sh::FastDDS::Server] Subscriber for topic 'ArmMotorsService_Reply' matched
[Integration Service][INFO] [is::sh::FastDDS] Server created for service 'ArmMotorsService', with request_type 'ArmMotors_Request' and reply_type 'ArmMotors_Response'
[Integration Service][INFO] [is::core::Config]  [dds SystemHandle] Produced a service provider for the service 'arm_motors', with request type 'ArmMotors_Request' and reply type 'ArmMotors_Response'.
[Integration Service][INFO] [is::sh::ROS2] Creating entity using the default QoS.
[Integration Service][INFO] [is::sh::ROS2] Created service client for service 'arm_motors' with type 'ap_custom_services/ArmMotors:response' on node '/is_ros2_node_16807'
[Integration Service][INFO] [is::core::Config] [ros2 SystemHandle] Produced a service client for the service 'arm_motors', with request type 'ArmMotors_Request' and reply type 'ArmMotors_Response'.
2023-07-08 16:15:11.918 [RTPS_READER_HISTORY Error] Change payload size of '12' bytes is larger than the history payload size of '8' bytes and cannot be resized. -> Function can_change_be_added_nts
```

Micro XRCE Client arming service:

```bash
$ ./ArmMotorsMicro 127.0.0.1 2019
Entities Created Successfully 
Request recieved for Arming 
Reply : Armed 
```

Micro XRCE Agent:

```bash
$ MicroXRCEAgent udp4 -p 2019 -v6
[1688829296.438878] info     | UDPv4AgentLinux.cpp | init                     | running...             | port: 2019
[1688829296.439526] info     | Root.cpp           | set_verbose_level        | logger setup           | verbose_level: 6
[1688829299.514055] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0x00000000, len: 24, data: 
0000: 80 00 00 00 00 01 10 00 58 52 43 45 01 00 01 0F CC CC DD DD 81 00 FC 01
[1688829299.514166] info     | Root.cpp           | create_client            | create                 | client_key: 0xCCCCDDDD, session_id: 0x81
[1688829299.514204] info     | SessionManager.hpp | establish_session        | session established    | client_key: 0xCCCCDDDD, address: 127.0.0.1:43263
[1688829299.514260] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 19, data: 
0000: 81 00 00 00 04 01 0B 00 00 00 58 52 43 45 01 00 01 0F 00
[1688829299.514337] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0xCCCCDDDD, len: 280, data: 
0000: 81 80 00 00 01 05 60 00 00 0A 00 11 01 02 00 42 51 00 00 00 3C 64 64 73 3E 3C 70 61 72 74 69 63
0020: 69 70 61 6E 74 3E 3C 72 74 70 73 3E 3C 6E 61 6D 65 3E 41 72 6D 4D 6F 74 6F 72 73 5F 53 65 72 76
0040: 65 72 3C 2F 6E 61 6D 65 3E 3C 2F 72 74 70 73 3E 3C 2F 70 61 72 74 69 63 69 70 61 6E 74 3E 3C 2F
0060: 64 64 73 3E 00 8B 00 00 01 05 AC 00 00 0B 00 18 08 02 00 00 9E 00 00 00 3C 64 64 73 3E 3C 72 65
0080: 70 6C 69 65 72 20 70 72 6F 66 69 6C 65 5F 6E 61 6D 65 3D 22 41 72 6D 4D 6F 74 6F 72 73 5F 52 65
00A0: 70 6C 69 65 72 22 73 65 72 76 69 63 65 5F 6E 61 6D 65 3D 22 41 72 6D 4D 6F 74 6F 72 73 53 65 72
00C0: 76 69 63 65 22 72 65 71 75 65 73 74 5F 74 79 70 65 3D 22 41 72 6D 4D 6F 74 6F 72 73 5F 52 65 71
00E0: 75 65 73 74 22 72 65 70 6C 79 5F 74 79 70 65 3D 22 41 72 6D 4D 6F 74 6F 72 73 5F 52 65 73 70 6F
0100: 6E 73 65 22 3E 3C 2F 72 65 70 6C 69 65 72 3E 3C 2F 64 64 73 3E 00 00 11
[1688829299.530461] info     | ProxyClient.cpp    | create_participant       | participant created    | client_key: 0xCCCCDDDD, participant_id: 0x001(1)
[1688829299.530528] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 14, data: 
0000: 81 80 00 00 05 01 06 00 00 0A 00 11 00 00
[1688829299.530736] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0xCCCCDDDD, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 01 00 00 00 80
[1688829299.532110] info     | ProxyClient.cpp    | create_replier           | replier created        | client_key: 0xCCCCDDDD, requester_id: 0x001(7), participant_id: 0x001(1)
[1688829299.532153] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 14, data: 
0000: 81 80 01 00 05 01 06 00 00 0B 00 18 00 00
[1688829299.532171] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 01 00 00 00 80
[1688829299.532221] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0xCCCCDDDD, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 02 00 00 00 80
[1688829299.532269] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0xCCCCDDDD, len: 24, data: 
0000: 81 80 01 00 08 01 10 00 00 0C 00 18 80 00 00 01 FF FF 00 00 00 00 00 00
[1688829299.532365] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 02 00 00 00 80
[1688829310.917022] debug    | Replier.cpp        | read_fn                  | [==>> DDS <<==]        | client_key: 0x00000001, len: 28, data: 
0000: 01 0F 7A 56 DE 01 9C E7 01 00 00 00 00 00 01 03 00 00 00 00 01 00 00 00 01 00 00 00
[1688829310.917083] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 40, data: 
0000: 81 80 02 00 09 01 20 00 00 0C 00 18 01 0F 7A 56 DE 01 9C E7 01 00 00 00 00 00 01 03 00 00 00 00
0020: 01 00 00 00 01 00 00 00
[1688829310.917152] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0xCCCCDDDD, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 03 00 00 00 80
[1688829311.918279] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0xCCCCDDDD, len: 44, data: 
0000: 81 80 02 00 07 01 24 00 00 0D 00 18 01 0F 7A 56 DE 01 9C E7 01 00 00 00 00 00 01 03 00 00 00 00
0020: 01 00 00 00 01 00 00 00 00 00 00 00
[1688829311.918446] debug    | Replier.cpp        | write                    | [** <<DDS>> **]        | client_key: 0x00000001, len: 32, data: 
0000: 01 0F 7A 56 DE 01 9C E7 01 00 00 00 00 00 01 03 00 00 00 00 01 00 00 00 01 00 00 00 00 00 00 00
[1688829311.918508] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0xCCCCDDDD, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 03 00 00 00 80
```

srmainwaring commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -433,6 +437,54 @@ void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, ui
+            .id = services[to_underlying(ServiceIndex::ARMING_MOTORS)].rep_id,
+            .type = UXR_REPLIER_ID
+        };
+        uint8_t reply_buffer[8] = {
```
```suggestion
        uint8_t reply_buffer[1] = {
```

Fixes [RTPS_READER_HISTORY Error].

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -433,6 +437,54 @@ void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, ui
+
+        // dummy code
+        // TODO : add arming/disarming checks here 
+        bool result = arm;
```
you should be able to call to AP_Arming like this:
 - AP::arming().arm(AP_Arming::Method::DDS)
 - AP::arming().disarm(AP_Arming::Method::DDS)

you will need to add DDS as an arming method in AP_Arming.h

arshPratap replied:
Done.. Thanks for this @tridge !

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -586,18 +641,56 @@ bool AP_DDS_Client::create()
+        //status requests
+        constexpr uint8_t nRequests = 1;
+        uint16_t requests[nRequests];
+        constexpr uint16_t requestTimeoutMs = (uint8_t) nRequests * maxTimeMsPerRequestMs;
```
why uint8_t cast?

arshPratap replied:
Thanks for this !

srmainwaring commented:
Hi @arshPratap, when including:

- https://github.com/arshPratap/ardupilot/pull/8
- https://github.com/arshPratap/ardupilot/pull/9

the commits to `Tools/ros2` should be split into separate commits from those to `AP_Tools`.

Might be easier if we break them out from this PR, get - https://github.com/arshPratap/ardupilot/pull/8 merged first and follow up with https://github.com/arshPratap/ardupilot/pull/9 after this.

srmainwaring commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -433,6 +437,54 @@ void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, ui
+        uint32_t* count_ptr = (uint32_t*) args;
+        (*count_ptr)++;
+
+        uxr_buffer_reply(uxr_session, reliable_out, replier_id, sample_id, reply_buffer, sizeof(reply_buffer));
```
```suggestion
        uxr_buffer_reply(uxr_session, reliable_out, replier_id, sample_id, reply_buffer, sizeof(result));
```

Correcting my earlier suggestion. The last arg should be the size of the serialised data, not the buffer size.

arshPratap replied:
Thanks a lot @srmainwaring ! Have made the requested changes

srmainwaring reviewed (commented):
@arshPratap pls change the prefix on the message in this commit to `AP_Arming:` and remove the changes to `AP_DDS_Client.cpp` and `AP_DDS_Client.h`. There should only be changes to files in the `AP_Arming` library.

srmainwaring commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -402,9 +408,8 @@ bool AP_DDS_Client::start(void)
 
 // read function triggered at every subscription callback
-void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, uint16_t request_id, uxrStreamId stream_id, struct ucdrBuffer* ub, uint16_t length, void* args)
+void AP_DDS_Client::on_topic ([[maybe_unused]] uxrSession* uxr_session, uxrObjectId object_id, [[maybe_unused]] uint16_t request_id, [[maybe_unused]] uxrStreamId stream_id, struct ucdrBuffer* ub, [[maybe_unused]] uint16_t length, void* args)
 {
-    (void) uxr_session; (void) object_id; (void) request_id; (void) stream_id; (void) length;
```
Revert [[maybe_unused]] as it is `--std=c++17`, which we are not currently using. @Ryanf55 do you agree?

srmainwaring commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -433,6 +438,55 @@ void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, ui
 
+#include "AP_DDS_Service_Table.h"
+
+void AP_DDS_Client::on_request([[maybe_unused]] uxrSession* uxr_session, uxrObjectId object_id, [[maybe_unused]] uint16_t request_id, SampleIdentity* sample_id, ucdrBuffer* ub, [[maybe_unused]] uint16_t length, void* args)
```
See comment above on `[[maybe_unused]]`.

arshPratap replied:
Done

srmainwaring commented on libraries/AP_DDS/README.md:
```diff
@@ -230,7 +230,45 @@ Next, follow the associated section for your chosen transport, and finally you c
+  - Get System Handles for [Fast-DDS](https://github.com/eProsima/FastDDS-SH)
+
+### Setup
+- Clone the [ardupilot_ros2](https://github.com/arshPratap/ardupilot_ros2) repo in your ROS 2 workspace
+- Switch to **ArmingClient** branch
+- Build the [ap_custom_services](https://github.com/arshPratap/ardupilot_ros2/tree/ArmingClient/ap_custom_services) package
+- Build the [ap_service_clients](https://github.com/arshPratap/ardupilot_ros2/tree/ArmingClient/ap_service_clients) package
+- In a new terminal, successfully source ROS 2 installation
+- Move to your ROS 2 workspace and source the installation (you should now see the above 2 pacakges when you run `ros2 pkg list`)
+- Move to your Integration Service workspace and use the following command to successfully link the ***ap_custom_services*** package with Integration Services `colcon build --cmake-args -DMIX_ROS_PACKAGES="ap_custom_services"`
```
Update this section to reflect that the message and service interfaces are defined in `ardupilot_msgs` (https://github.com/ArduPilot/ardupilot/pull/24277).

arshPratap replied:
Done

arshPratap commented:
@srmainwaring thanks for the reviews..have added the said requested changes accordingly.

srmainwaring commented on libraries/AP_DDS/Is-Config/Arm_Motors_DDS_IS_config.yaml:
```diff
@@ -0,0 +1,36 @@
+            dds: {
+                topic: ArmMotorsService,
+            },
+            ros2: {
+                request_type: "ap_custom_services/ArmMotors:request",
+                reply_type: "ap_custom_services/ArmMotors:response"
+            }
```
```suggestion
            ros2: {
                request_type: "ardupilot_msgs/ArmMotors:request",
                reply_type: "ardupilot_msgs/ArmMotors:response"
            }
```

arshPratap replied:
Done

srmainwaring commented on libraries/AP_DDS/README.md:
```diff
@@ -230,7 +230,32 @@ Next, follow the associated section for your chosen transport, and finally you c
+- Run SITL with the following command `sim_vehicle.py -v ArduPlane -DG --console --enable-dds`
+
+### Terminal 4 (ROS 2 Client)
+- Run the following command : `ros2 service call /arm_motors ap_custom_services/srv/ArmMotors "{arm: True}"`
```
```suggestion
- Run the following command : `ros2 service call /arm_motors ardupilot_msgs/srv/ArmMotors "{arm: True}"`
```

arshPratap replied:
Done

arshPratap commented:
@srmainwaring thanks again ! Have worked in the said changes

srmainwaring commented:
> Have worked in the said changes

Thanks @arshPratap, LGTM.

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -402,26 +404,30 @@ bool AP_DDS_Client::start(void)
-        (*count_ptr)++;
+        subscribe_sample_count++;
         if (topic.axes_size >= 4) {
             GCS_SEND_TEXT(MAV_SEVERITY_INFO,"Received sensor_msgs/Joy: %f, %f, %f, %f",
```
these messages will be very annoying. You could have a DDS_OPTIONS parameter with a bit which enables debug. Otherwise add a debug() macro. See AP_BLHeli.cpp for a debug macro example, it would default off

Ryanf55 replied:
These were planned to be removed once the implementation of RC input is complete.

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -430,7 +436,64 @@ void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, ui
+            GCS_SEND_TEXT(MAV_SEVERITY_INFO,"DDS Client: Request for arming recieved");
+            result = AP::arming().arm(AP_Arming::Method::DDS);
+        } else {
+            result = !AP::arming().disarm(AP_Arming::Method::DDS);
```
result in the reply should be true if we successfully disarmed

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -430,7 +436,64 @@ void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, ui
+        request_sample_count++;
+
+        uxr_buffer_reply(uxr_session, reliable_out, replier_id, sample_id, reply_buffer, sizeof(result));
+        if (result == true) {
```
see above comments on debug code

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -586,18 +652,56 @@ bool AP_DDS_Client::create()
 
             if (!uxr_run_session_until_all_status(&session, requestTimeoutMs, requests, status, nRequests)) {
-                GCS_SEND_TEXT(MAV_SEVERITY_ERROR,"XRCE Client: Topic/Sub/Reader session request failure for index '%d'",(int)i);
+                GCS_SEND_TEXT(MAV_SEVERITY_ERROR,"XRCE Client: Topic/Sub/Reader session request failure for index '%zu'",i);
```
what is %zu ?

tridge replied:
also see comments on debug code

Ryanf55 replied:
%zu is for size_t type. 

https://cplusplus.com/reference/cstdio/printf/

The `ARRAY_SIZE` macro calls `sizeof()` which returns a `size_t`. Since ArduPilot doesn't support containers with range-based for loops, we must manually set the type.

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -586,18 +652,56 @@ bool AP_DDS_Client::create()
+
+    // ROS-2 Service : else case for service requests
+
+    for (size_t i = 0; i < ARRAY_SIZE(services); i++) {
```
why a size_t? maybe a uint16_t ?

Ryanf55 replied:
The `ARRAY_SIZE` macro calls `sizeof()` which returns a `size_t`. Since ArduPilot doesn't support containers with range-based for loops, we must manually set the type, so I chose one that matched what was returned by ARRAY_SIZE.

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -586,18 +652,56 @@ bool AP_DDS_Client::create()
+    for (size_t i = 0; i < ARRAY_SIZE(services); i++) {
+
+        //status requests
+        constexpr uint8_t nRequests = 1;
```
why have the arrays when it is always 1?

Ryanf55 replied:
Arsh is probably mirroring the other structure established where publishers and subscribers had multiple requests. Are you just concerned about flash overhead here? This code only runs during initialization

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -586,18 +652,56 @@ bool AP_DDS_Client::create()
+
+            requests[0] = replier_req_id;
+
+            if (!uxr_run_session_until_all_status(&session, requestTimeoutMs, requests, status, nRequests)) {
```
```
uxr_run_session_until_all_status(&session, requestTimeoutMs, &requests, &status, 1)
```

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -586,18 +652,56 @@ bool AP_DDS_Client::create()
+
+            if (!uxr_run_session_until_all_status(&session, requestTimeoutMs, requests, status, nRequests)) {
+                GCS_SEND_TEXT(MAV_SEVERITY_ERROR,"XRCE Client: Service/Replier session request failure for index '%zu'",i);
+                for (uint8_t s = 0 ; s < nRequests; s++) {
```
only ever 1 in this code, no loop

tridge reviewed (changes requested):
please cleanup debug code, and remove unnecessary loops
thanks!

tridge commented:
@rmackay9 should we rename Method::MAVLINK to Method::GCS so the checks in many places that have special behaviour for mavlink arm/disarm are duplicated for DDS without extra code? Or we could have a AP_Arming::is_GCS_method() that returns true for both DDS and MAVLink?

tridge commented:
for the AP_Arming::Method::MAVLINK checks, adding a static method on AP_Arming called method_is_GCS() would make sense
bool AP_Arming::method_is_GCS(Method method);

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -606,18 +680,53 @@ bool AP_DDS_Client::create()
+                debug("XRCE Client: Topic/Sub/Reader session request failure for index '%u'",i);
                 for (uint8_t s = 0 ; s < nRequests; s++) {
-                    GCS_SEND_TEXT(MAV_SEVERITY_ERROR,"XRCE Client: Status '%d' result '%u'", s, status[s]);
+                    debug("XRCE Client: Status '%d' result '%u'", s, status[s]);
```
Just curious - what is the reason to make these optional debug messages, all at the same level?

arshPratap replied:
@Ryanf55 tridge requested for the session failure messages for topics and services to be made as debug messages .. so to keep things in line I made the status messages as debug

Ryanf55 replied:
Ok, I'll follow up with Tridge - Errors would now be silently swallowed, but other libraries are allowed to call `GCS_SEND_TEXT(MAV_SEVERITY_ERROR` during initialization.

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -440,19 +456,74 @@ void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, ui
+            .id = services[to_underlying(ServiceIndex::ARMING_MOTORS)].rep_id,
+            .type = UXR_REPLIER_ID
+        };
+        uint8_t reply_buffer[8] {};
```
add comment explaining why 8

tridge commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -440,19 +456,74 @@ void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, ui
+
+        request_sample_count++;
+
+        uxr_buffer_reply(uxr_session, reliable_out, replier_id, sample_id, reply_buffer, sizeof(result));
```
should sizeof(result) be reply_ub.offset?
maybe use ucdr_buffer_length(reply_buffer) ?

arshPratap commented:
@tridge thanks for the review ..have made the changes accordingly

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.h:
```diff
@@ -51,8 +52,6 @@ class AP_DDS_Client
     // input and output stream
     uint8_t *input_reliable_stream;
     uint8_t *output_reliable_stream;
-    uxrStreamId reliable_in;
```
Why were these moved away from the pointers to a different part of the header?

arshPratap replied:
done resolved

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -606,18 +684,53 @@ bool AP_DDS_Client::create()
+    for (uint16_t i = 0; i < ARRAY_SIZE(services); i++) {
+
+        //status requests
+        uint16_t request;
```
This has larger scope than needed. It's just used to create a copy of data once. Can you try deleting it?

arshPratap replied:
Resolved

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -606,18 +684,53 @@ bool AP_DDS_Client::create()
+        //status requests
+        uint16_t request;
+        constexpr uint16_t requestTimeoutMs = maxTimeMsPerRequestMs;
+        uint8_t status;
```
Decrease the scope of this to where it's used.

arshPratap replied:
done .. resolved

Ryanf55 commented on libraries/AP_DDS/README.md:
```diff
@@ -231,7 +231,32 @@ Next, follow the associated section for your chosen transport, and finally you c
 
+### Prerequisites
+- Install and setup [Micro-XRCE Agent](https://micro-xrce-dds.docs.eprosima.com/en/latest/installation.html#installing-the-agent-standalone)
+- Install and setup [Integration Services](https://micro-xrce-dds.docs.eprosima.com/en/latest/installation.html#installing-the-client-standalone) (it would be good to have a seperate workspace for this)
```
Is this the right link? It's not the integration service.

arshPratap replied:
fixed the link..thanks for pointing out

Ryanf55 reviewed (changes requested):
I was unable to complete testing - the README seems to have an incorrect link.

Ryanf55 commented on libraries/AP_DDS/README.md:
```diff
@@ -10,21 +10,21 @@ title: UDP Loopback
-      veh[sim_vehicle.py] <--> xrceClient[EProsima Micro XRCE DDS Client]
-      xrceClient <--> port1[udp:2019]
-    end
+subgraph Ardupilot SITL
```
Did you accidentally re-format the whole README? This indent is now messed up.

arshPratap replied:
I ran the astyle on the readme file..no worries..have reverted the change

Ryanf55 replied:
Thanks, yea `astyle` is only used for the C++ code.

Ryanf55 commented on libraries/AP_DDS/README.md:
```diff
@@ -231,6 +235,31 @@ Next, follow the associated section for your chosen transport, and finally you c
+
+### Prerequisites
+- Install and setup [Micro-XRCE Agent](https://micro-xrce-dds.docs.eprosima.com/en/latest/installation.html#installing-the-agent-standalone)
+- Install and setup [Integration Services](https://integration-service.docs.eprosima.com/en/latest/installation_manual/installation.html) (it would be good to have a seperate workspace for this)
```
```suggestion
- Install and setup [Integration Services](https://integration-service.docs.eprosima.com/en/latest/installation_manual/installation.html) (it would be good to have a separate workspace for this)
```
Spelling

arshPratap replied:
Resolved

arshPratap commented on libraries/AP_DDS/dds_xrce_profile.xml:
```diff
@@ -253,4 +253,9 @@
     </topic>
   </data_reader>
+  <replier profile_name = "ArmMotors_Replier" 
+           service_name = "ArmMotorsService" 
```
Added `ros2 service list` as the command ..since integration service creates another node .. but otherwise added the service description

arshPratap commented:
@Ryanf55 thanks for the review
Have made the changes as required

srmainwaring commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -440,19 +459,75 @@ void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, ui
+        }
+
+        if (arm) {
+            debug("Request for arming recieved");
```
```suggestion
            debug("Request for arming received");
```

arshPratap replied:
resolved

srmainwaring commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -440,19 +459,75 @@ void AP_DDS_Client::on_topic (uxrSession* uxr_session, uxrObjectId object_id, ui
+            debug("Request for arming recieved");
+            result = AP::arming().arm(AP_Arming::Method::DDS);
+        } else {
+            debug("Request for disarming recieved");
```
```suggestion
            debug("Request for disarming received");
```

arshPratap replied:
resolved

srmainwaring commented on libraries/AP_DDS/README.md:
```diff
@@ -231,7 +234,32 @@ Next, follow the associated section for your chosen transport, and finally you c
+- Install and setup [Integration Services](https://integration-service.docs.eprosima.com/en/latest/installation_manual/installation.html) (it would be good to have a separate workspace for this)
+  - Get System Handles for [ROS 2](https://github.com/eProsima/ROS2-SH)
+  - Get System Handles for [Fast-DDS](https://github.com/eProsima/FastDDS-SH)
+
```
Add a note that the `ROS2-SH` packages need to be built with additional flags:

```bash
--cmake-args -DMIX_ROS_PACKAGES="example_interfaces ardupilot_msgs"
```

arshPratap replied:
Resolved

srmainwaring commented:
Hi @arshPratap - thanks for getting all the changes in. I have run a clean build and test of the most recent version and appears to be working. Tested the following:

- Build - OK.
    - Added request to update README to include extra cmake flags for the is-ros-mix-generator SH.
- SITL - OK
    - Run using the UDP `launch.py`.
- Integration Service - OK
    - Need additional flags to locate the .mix files and libs on macOS. Not required on Ubuntu.
- DDS_DEBUG param - OK
- Service call - OK
    - Vehicle arms / disarms.
    - A second call to arm should return false if the vehicle is already armed, however I think this is an upstream issue as the service call is directly returning `AP::arming().arm(AP_Arming::Method::DDS)`.

Ryanf55 commented:
With the testing of @srmainwaring , I'm happy to let this in and resolve the integration service compilation issues on my side separately. @arshPratap Please fix the merge conflicts, and we can merge it. Many thanks.

arshPratap commented:
@srmainwaring @Ryanf55 thanks for the review and testing .. I have made the requested changes  and have pushed in the code

Ryanf55 commented:
Please run pre-commit, it is failing on `libraries/AP_DDS/dds_xrce_profile.xml`

arshPratap commented:
> Please run pre-commit, it is failing on `libraries/AP_DDS/dds_xrce_profile.xml`. Ready to merge otherwise.
> 
> ```shell
> cd /your/path/to/ardupilot
> python3 -m pip install pre-commit
> pre-commit install
> pre-commit run --files libraries/AP_DDS/* -v
> ```

thanks @Ryanf55 ... have applied  the changes

Ryanf55 reviewed (changes requested):
As mentioned in chat, I am strongly opposed to the modified logging method that is globally applied to all the logging. I would be happy to accept the PR without it.

arshPratap commented:
> As mentioned in chat, I am strongly opposed to the modified logging method that is globally applied to all the logging. I would be happy to accept the PR without it. If it's not too work, split those changes to another PR, and this can go in. Changing the logging methodology to hide errors is not part of ROS 2 service support.

@tridge is it alright if I split the logging changes to another PR  as @Ryanf55 suggests  ? Or should it be in this PR

tridge commented:
@Ryanf55 I'm fine with the debug being a separate PR. It should probably take an enum level. Main reason for changing is right now we do SEND_TEXT on every message, which would be complete impractical while flying (would flood the radio link)

Ryanf55 commented:
> @Ryanf55 I'm fine with the debug being a separate PR. It should probably take an enum level. Main reason for changing is right now we do SEND_TEXT on every message, which would be complete impractical while flying (would flood the radio link). For example, on every joystick pkt

Thanks for the explanation. Totally on board; it wasn't the intent to leave those once the interfaces were developed. We can figure better plan in the next PR.

Ryanf55 commented on libraries/AP_DDS/AP_DDS_Client.cpp:
```diff
@@ -593,14 +664,14 @@ bool AP_DDS_Client::create()
 
             if (!uxr_run_session_until_all_status(&session, requestTimeoutMs, requests, status, nRequests)) {
-                GCS_SEND_TEXT(MAV_SEVERITY_ERROR,"XRCE Client: Topic/Pub/Writer session request failure for index 'TODO'");
+                GCS_SEND_TEXT(MAV_SEVERITY_INFO,"XRCE Client: Topic/Pub/Writer session request failure for index '%u'",i);
```
This should be an error.

arshPratap replied:
resolved

arshPratap commented:
thanks @Ryanf55  for the review..have made the changes as requested
